streams/src/main/java/org/apache/kafka/streams/state/internals/WrappedStateStore.java

```diff
@@ -19,81 +19,66 @@
 import org.apache.kafka.streams.errors.InvalidStateStoreException;
 import org.apache.kafka.streams.processor.ProcessorContext;
 import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.state.TimestampedBytesStore;
 
 /**
  * A storage engine wrapper for utilities like logging, caching, and metering.
  */
-public interface WrappedStateStore extends StateStore {
-
-    /**
-     * Return the inner most storage engine
-     *
-     * @return wrapped inner storage engine
-     */
-    StateStore inner();
-
-    /**
-     * Return the state store this store directly wraps
-     * @return the state store this store directly wraps
-     */
-    StateStore wrappedStore();
-
-    abstract class AbstractStateStore implements WrappedStateStore {
-        final StateStore innerState;
-
-        protected AbstractStateStore(final StateStore inner) {
-            this.innerState = inner;
+public class WrappedStateStore<S extends StateStore> implements StateStore {
+    public static boolean isTimestamped(StateStore stateStore) {
+        if (stateStore instanceof TimestampedBytesStore) {
+            return true;
+        } else if (stateStore instanceof WrappedStateStore) {
+            return isTimestamped(((WrappedStateStore) stateStore).wrappedStore());
```

Comment:
I think, we should first check for this condition, because we should only check the most inner store -- if an wrapping store would (be mistake) implement `TimestampedBytesStore`, we would return `true` even if the most inner store does not -- this would be incorrect.

Reply:
Interesting thought.

Let's say there's a wrapper that advertises that it's a TimestampedBytesStore. This means it is telling Streams that it expects to receive serialized data in the new (with-timestamp) format.

Why does it do this? Maybe it's a translation layer for other stores? In which case, is it correct for Streams to second-guess the implementation and break its own contract by ignoring the marker interface and delivering non-timestamped binary data?

Or maybe, as you suggest, it's just a bug. Is it best for Streams to try and work around a future bug in a state store implementation by reasoning about relative roles of wrapper stores and bytes stores?

I'm not sure... It sounds like it would be more future-proof to just take the store's interface at face value.

Reply:
> Why does it do this? Maybe it's a translation layer for other stores? In which case, is it correct for Streams to second-guess the implementation and break its own contract by ignoring the marker interface and delivering non-timestamped binary data?

I don't think that would work. Note, on restore, we always get the most inner store and would not call this "translation layer wrapper store" (and thus it would break as we would insert our converter and hand timestamped-bytes to the store that does not understand them). If one want to implement a translation wrapper like this, she need to "hide" it from Kafka Streams and not implement `WrappingStore` (ie, the translation wrapper must be the most inner store).

Reply:
Huh... How do we get the most inner store on restore?

Reply:
`WrappedStore` should have a method `root()` (or similar name) that returns the most inner store.

Reply:
As far as I can tell, this method would actually never be needed in the existing code base.

I agree with the naming of it (and the location in the code), but can we defer adding it until it's needed?

To address the earlier question, I believe that the state restoration logic just calls whatever restore listener has been registered. It turns out that the root store is the one to register the listener, but this doesn't require actually unwrapping anything (since it is done within context).